descqa/NumberDensityVersusRedshift.py

```diff
@@ -0,0 +1,322 @@
+                summary = self.make_subplot(meanz,catalog_data[self.z][mask],n,self.summary_ax[nrow,ncol],z0,z0err,catalog_color,validation_color,catalog_label,validation_label)
+            else:
+                summary = self.make_subplot(meanz,catalog_data[self.z][mask],n,self.summary_ax[nrow,ncol],0.,0.,catalog_color,validation_color,catalog_label,validation_label)
+
+        #make empty subplots invisible 
+        if (len(self.mag_lo)%self.ncolumns!=0): #check for empty subplots
+                ax[self.nrows-1,self.ncolumns-1].set_visible(False)   #assumes ncolumns=2
+                
+        #save results for catalog and validation data in txt files
+        for filename, dtype, comment, info in zip([filelabel,self.observation],['y','fit'],[filtername,''],['total','']):
+            with open(os.path.join(base_output_dir,'Nvsz_'+filename+'.txt'),'ab') as f_handle:     #open file in append mode
+                 for key in results.keys():
+                     xinfo = results[key][info] if len(info)>0 else ''
+                     self.save_quantities(dtype, results[key], f_handle,comment=' '.join([comment,key,xinfo]))
+
+        if(self.first_pass): #turn off validation data plot in summary for remaining catalogs
+            self.first_pass = False
+
+        #save figure
+        plt.savefig(os.path.join(base_output_dir, 'Nvsz_'+filelabel+'.png'))
+        plt.close()
+        return TestResult(0, passed=True)
+
+    def make_subplot(self,meanz,catalog_data,nplot,f,z0,z0err,catalog_color,validation_color,catalog_label,validation_label):
+
+        results={}
+        results['meanz']=meanz
+        if(nplot%self.ncolumns==0):  #1st column
+            f.set_ylabel('$'+self.yaxis+'$',size=fsize)
+
+        if(nplot+1 <= self.nplots-self.ncolumns):  #x scales for last ncol plots only
+            #print "noticks",nplot
+            for axlabel in f.get_xticklabels():
+                axlabel.set_visible(False)
+                    
+                #prevent overlapping yaxis labels
+                f.yaxis.get_major_ticks()[0].label1.set_visible(False)
+        else:
+            f.set_xlabel('$z$',size=fsize)
+            for axlabel in f.get_xticklabels():
+                axlabel.set_visible(True)
+
+        #plot catalog data if available
+        if(len(catalog_data)>0):
+            y,binEdges,_=f.hist(catalog_data,bins=self.zbins,label=catalog_label,color=catalog_color,lw=lw2,normed=self.normed,histtype='step')
+            results['y']=y
+
+        #plot validation data if available
+        if(z0>0.):
+            ndata=meanz**2*np.exp(-meanz/z0)
+            if(self.normed):
+                norm=self.nz_norm(self.zhi,z0)-self.nz_norm(self.zlo,z0)
+                f.plot(meanz,ndata/norm,label=validation_label,ls='--',color=validation_color,lw=lw2)
+                results['fit']=ndata/norm
+            else:
+                print('oops, needs work!')
```

Comment:
This is fine but I wonder if it should raise `NotImplementedError`?

Reply:
Sure, can you remind me how to raise the error?